File: app-java/src/main/java/com/google/sample/cast/refplayer/queue/ui/QueueListViewActivity.java
/*
 * Copyright 2022 Google LLC. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.sample.cast.refplayer.queue.ui;

import com.google.android.gms.cast.MediaQueueItem;
import com.google.android.gms.cast.MediaStatus;
import com.google.android.gms.cast.framework.CastButtonFactory;
import com.google.android.gms.cast.framework.CastContext;
import com.google.android.gms.cast.framework.CastSession;
import com.google.android.gms.cast.framework.SessionManagerListener;
import com.google.android.gms.cast.framework.media.RemoteMediaClient;
import com.google.sample.cast.refplayer.R;
import com.google.sample.cast.refplayer.queue.QueueDataProvider;
import com.google.sample.cast.refplayer.settings.CastPreference;

import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import android.util.Log;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

/**
 * An activity to show the queue list
 */
public class QueueListViewActivity extends AppCompatActivity {

    private static final String FRAGMENT_LIST_VIEW = "list view";
    private static final String TAG = "QueueListViewActivity";

    private final RemoteMediaClient.Callback mRemoteMediaClientCallback =
            new MyRemoteMediaClientCallback();
    private final SessionManagerListener<CastSession> mSessionManagerListener =
            new MySessionManagerListener();
    private CastContext mCastContext;
    private RemoteMediaClient mRemoteMediaClient;
    private View mEmptyView;
    private Executor localExecutor = Executors.newSingleThreadExecutor();

    private class MySessionManagerListener implements SessionManagerListener<CastSession> {

        @Override
        public void onSessionEnded(CastSession session, int error) {
            if (mRemoteMediaClient != null) {
                mRemoteMediaClient.unregisterCallback(mRemoteMediaClientCallback);
            }
            mRemoteMediaClient = null;
            mEmptyView.setVisibility(View.VISIBLE);
        }

        @Override
        public void onSessionResumed(CastSession session, boolean wasSuspended) {
            mRemoteMediaClient = getRemoteMediaClient();
            if (mRemoteMediaClient != null) {
                mRemoteMediaClient.registerCallback(mRemoteMediaClientCallback);
            }
        }

        @Override
        public void onSessionStarted(CastSession session, String sessionId) {
            mRemoteMediaClient = getRemoteMediaClient();
            if (mRemoteMediaClient != null) {
                mRemoteMediaClient.registerCallback(mRemoteMediaClientCallback);
            }
        }

        @Override
        public void onSessionStarting(CastSession session) {
        }

        @Override
        public void onSessionStartFailed(CastSession session, int error) {
        }

        @Override
        public void onSessionEnding(CastSession session) {
        }

        @Override
        public void onSessionResuming(CastSession session, String sessionId) {
        }

        @Override
        public void onSessionResumeFailed(CastSession session, int error) {
        }

        @Override
        public void onSessionSuspended(CastSession session, int reason) {
            if (mRemoteMediaClient != null) {
                mRemoteMediaClient.unregisterCallback(mRemoteMediaClientCallback);
            }
            mRemoteMediaClient = null;
        }
    }

    private class MyRemoteMediaClientCallback extends RemoteMediaClient.Callback {
        @Override
        public void onStatusUpdated() {
            updateMediaQueue();
        }

        @Override
        public void onQueueStatusUpdated() {
            updateMediaQueue();
        }

        private void updateMediaQueue() {
            MediaStatus mediaStatus = mRemoteMediaClient.getMediaStatus();
            List<MediaQueueItem> queueItems =
                    (mediaStatus == null) ? null : mediaStatus.getQueueItems();
            if (queueItems == null || queueItems.isEmpty()) {
                mEmptyView.setVisibility(View.VISIBLE);
            } else {
                mEmptyView.setVisibility(View.GONE);
            }
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.queue_activity);
        Log.d(TAG, "onCreate() was called");

        if (savedInstanceState == null) {
            getSupportFragmentManager().beginTransaction()
                    .add(R.id.container, new QueueListViewFragment(), FRAGMENT_LIST_VIEW)
                    .commit();
        }
        setupActionBar();
        mEmptyView = findViewById(R.id.empty);
        mCastContext = CastContext.getSharedInstance(this,localExecutor).getResult();
    }


    private void setupActionBar() {
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        toolbar.setTitle(R.string.queue_list);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    @Override
    protected void onPause() {
        if (mRemoteMediaClient != null) {
            mRemoteMediaClient.unregisterCallback(mRemoteMediaClientCallback);
        }
        mCastContext.getSessionManager().removeSessionManagerListener(
                mSessionManagerListener, CastSession.class);
        super.onPause();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.queue_menu, menu);
        CastButtonFactory.setUpMediaRouteButton(getApplicationContext(), menu,
                R.id.media_route_menu_item);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_settings) {
            startActivity(new Intent(QueueListViewActivity.this, CastPreference.class));
        } else if (id == R.id.action_clear_queue) {
            QueueDataProvider.getInstance(getApplicationContext()).removeAll();
        } else if (id == android.R.id.home) {
            finish();
        }
        return true;
    }

    @Override
    public boolean dispatchKeyEvent(@NonNull KeyEvent event) {
        return mCastContext.onDispatchVolumeKeyEventBeforeJellyBean(event)
                || super.dispatchKeyEvent(event);
    }

    @Override
    protected void onResume() {
        mCastContext.getSessionManager().addSessionManagerListener(
                mSessionManagerListener, CastSession.class);
        if (mRemoteMediaClient == null) {
            mRemoteMediaClient = getRemoteMediaClient();
        }
        if (mRemoteMediaClient != null) {
            mRemoteMediaClient.registerCallback(mRemoteMediaClientCallback);
            MediaStatus mediaStatus = mRemoteMediaClient.getMediaStatus();
            List<MediaQueueItem> queueItems =
                    (mediaStatus == null) ? null : mediaStatus.getQueueItems();
            if (queueItems != null && !queueItems.isEmpty()) {
                mEmptyView.setVisibility(View.GONE);
            }
        }
        super.onResume();
    }

    private RemoteMediaClient getRemoteMediaClient() {
        CastSession castSession = mCastContext.getSessionManager().getCurrentCastSession();
        return (castSession != null && castSession.isConnected())
                ? castSession.getRemoteMediaClient() : null;
    }
}

File: app-kotlin/src/main/kotlin/com/google/sample/cast/refplayer/queue/ui/QueueListViewActivity.kt
/*
 * Copyright 2022 Google LLC. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.sample.cast.refplayer.queue.ui

import android.util.Log
import android.view.View
import com.google.sample.cast.refplayer.R
import com.google.android.gms.cast.framework.CastSession
import com.google.android.gms.cast.framework.CastContext
import com.google.android.gms.cast.framework.SessionManagerListener
import android.os.Bundle
import android.content.Intent
import android.view.Menu
import com.google.android.gms.cast.framework.CastButtonFactory
import androidx.appcompat.app.AppCompatActivity
import android.view.MenuItem
import com.google.android.gms.cast.framework.media.RemoteMediaClient
import android.view.KeyEvent
import androidx.appcompat.widget.Toolbar
import com.google.sample.cast.refplayer.queue.QueueDataProvider
import com.google.sample.cast.refplayer.settings.CastPreference
import java.util.concurrent.Executor
import java.util.concurrent.Executors

/**
 * An activity to show the queue list
 */
class QueueListViewActivity : AppCompatActivity() {
    private val mRemoteMediaClientCallback: RemoteMediaClient.Callback =
        MyRemoteMediaClientCallback()
    private val mSessionManagerListener: SessionManagerListener<CastSession> =
        MySessionManagerListener()
    private var mCastContext: CastContext? = null
    private val  castExecutor: Executor = Executors.newSingleThreadExecutor();
    private var mRemoteMediaClient: RemoteMediaClient? = null
    private var mEmptyView: View? = null

    private inner class MySessionManagerListener : SessionManagerListener<CastSession> {
        override fun onSessionEnded(session: CastSession, error: Int) {
            if (mRemoteMediaClient != null) {
                mRemoteMediaClient!!.unregisterCallback(mRemoteMediaClientCallback)
            }
            mRemoteMediaClient = null
            mEmptyView!!.visibility = View.VISIBLE
        }

        override fun onSessionResumed(session: CastSession, wasSuspended: Boolean) {
            mRemoteMediaClient = remoteMediaClient
            if (mRemoteMediaClient != null) {
                mRemoteMediaClient!!.registerCallback(mRemoteMediaClientCallback)
            }
        }

        override fun onSessionStarted(session: CastSession, sessionId: String) {
            mRemoteMediaClient = remoteMediaClient
            if (mRemoteMediaClient != null) {
                mRemoteMediaClient!!.registerCallback(mRemoteMediaClientCallback)
            }
        }

        override fun onSessionStarting(session: CastSession) {}
        override fun onSessionStartFailed(session: CastSession, error: Int) {}
        override fun onSessionEnding(session: CastSession) {}
        override fun onSessionResuming(session: CastSession, sessionId: String) {}
        override fun onSessionResumeFailed(session: CastSession, error: Int) {}
        override fun onSessionSuspended(session: CastSession, reason: Int) {
            if (mRemoteMediaClient != null) {
                mRemoteMediaClient!!.unregisterCallback(mRemoteMediaClientCallback)
            }
            mRemoteMediaClient = null
        }
    }

    private inner class MyRemoteMediaClientCallback : RemoteMediaClient.Callback() {
        override fun onStatusUpdated() {
            updateMediaQueue()
        }

        override fun onQueueStatusUpdated() {
            updateMediaQueue()
        }

        private fun updateMediaQueue() {
            val mediaStatus = mRemoteMediaClient!!.mediaStatus
            val queueItems = mediaStatus?.queueItems
            if (queueItems == null || queueItems.isEmpty()) {
                mEmptyView!!.visibility = View.VISIBLE
            } else {
                mEmptyView!!.visibility = View.GONE
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.queue_activity)
        Log.d(TAG, "onCreate() was called")
        if (savedInstanceState == null) {
            supportFragmentManager.beginTransaction()
                .add(R.id.container, QueueListViewFragment(), FRAGMENT_LIST_VIEW)
                .commit()
        }
        setupActionBar()
        mEmptyView = findViewById(R.id.empty)
        mCastContext = CastContext.getSharedInstance(this,castExecutor).result
    }

    private fun setupActionBar() {
        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        toolbar.setTitle(R.string.queue_list)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    }

    override fun onPause() {
        if (mRemoteMediaClient != null) {
            mRemoteMediaClient!!.unregisterCallback(mRemoteMediaClientCallback)
        }
        mCastContext!!.sessionManager.removeSessionManagerListener(
            mSessionManagerListener, CastSession::class.java
        )
        super.onPause()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.queue_menu, menu)
        CastButtonFactory.setUpMediaRouteButton(
            applicationContext, menu,
            R.id.media_route_menu_item
        )
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> startActivity(
                Intent(
                    this@QueueListViewActivity,
                    CastPreference::class.java
                )
            )
            R.id.action_clear_queue -> QueueDataProvider.Companion.getInstance(
                applicationContext
            )!!.removeAll()
            android.R.id.home -> finish()
        }
        return true
    }

    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
        return (mCastContext!!.onDispatchVolumeKeyEventBeforeJellyBean(event)
                || super.dispatchKeyEvent(event))
    }

    override fun onResume() {
        mCastContext!!.sessionManager.addSessionManagerListener(
            mSessionManagerListener, CastSession::class.java
        )
        if (mRemoteMediaClient == null) {
            mRemoteMediaClient = remoteMediaClient
        }
        if (mRemoteMediaClient != null) {
            mRemoteMediaClient!!.registerCallback(mRemoteMediaClientCallback)
            val mediaStatus = mRemoteMediaClient!!.mediaStatus
            val queueItems = mediaStatus?.queueItems
            if (queueItems != null && queueItems.isNotEmpty()) {
                mEmptyView!!.visibility = View.GONE
            }
        }
        super.onResume()
    }

    private val remoteMediaClient: RemoteMediaClient?
        private get() {
            val castSession = mCastContext!!.sessionManager.currentCastSession
            return if (castSession != null && castSession.isConnected) castSession.remoteMediaClient else null
        }

    companion object {
        private const val FRAGMENT_LIST_VIEW = "list view"
        private const val TAG = "QueueListViewActivity"
    }
}